DHT/dht.c

```diff
@@ -900,10 +913,9 @@ dhtElement *dhtEnterElement(HashTable *ht, dhtConstValue key, dhtConstValue data
     return dhtNilElement;
   }
 
-  DataV = data==0 ? 0 : (ht->procs.DupData)(data);
-  if (data!=0 && DataV==0)
+  if ((ht->procs.DupData)(data, &DataV))
   {
-    (ht->procs.FreeKey)((dhtValue)KeyV);
+    (ht->procs.FreeKeyValue)(KeyV.value);
     TraceText("data duplication failed\n");
```

Comment:
Hmm.
Before, we signaled "data duplication failed" if DupData returned 0.
Now, we signal it if it doesn't.
Can this be correct?

Reply:
Analogously to above, `DupData` is specified to return 0 on success and nonzero on failure, so we're still checking the analogous condition.&nbsp; What's special here was that previously we allowed for the original `data` to be `NULL` without signaling an error, hence I had to allow the same here.&nbsp; To make that work I had to allow `DupData` to simply copy a `NULL` pointer and return 0.